### PR TITLE
Fix/store regular loop in tree

### PR DESCRIPTION
**Pull request by pavlo-hilei**
Store regular (not attributed) loops in a loop tree. This is required for some edge cases, like mentioned in https://github.com/libocca/occa-transpiler/issues/150

### CODE REVIEW

File: lib/core/ast_processors/okl_sema_processor/processor.cpp [hunk 1]
```diff
@@ -7,11 +7,44 @@
...
+        }
+    }
+    return {};
+}
+
+HandleResult semaDefaultPost(const clang::Attr*,
+                             const clang::Stmt& stmt,
+                             OklSemaCtx& sema,
+                             SessionStage& stage) {
+    switch (stmt.getStmtClass()) {
```
Comment (kchabSS) [inline lib/core/ast_processors/okl_sema_processor/processor.cpp:36]: This one is better be simplified to `if (auto forStmt = dyn_cast_or_null<ForStmt>(&stmt))`

File: lib/core/ast_processors/okl_sema_processor/processor.cpp [hunk 2]
```diff
@@ -7,11 +7,44 @@
...
 namespace {
 using namespace clang;
 using namespace oklt;
 
+namespace {
+HandleResult semaDefaultPre(const clang::Attr*,
+                            const clang::Stmt& stmt,
+                            OklSemaCtx& sema,
+                            SessionStage& stage) {
+    switch (stmt.getStmtClass()) {
```
Comment (kchabSS) [inline lib/core/ast_processors/okl_sema_processor/processor.cpp:21]: This one is better be simplified to `if (auto forStmt = dyn_cast_or_null<ForStmt>(&stmt))`

File: lib/core/sema/okl_sema_info.h [hunk 3]
```diff
@@ -106,6 +106,7 @@ struct OklKernelInfo {
     explicit OklKernelInfo(const clang::FunctionDecl& decl)
         : decl(std::ref(decl)){};
     const std::reference_wrapper<const clang::FunctionDecl> decl;
-    std::list<OklLoopInfo> children = {};
+    std::list<OklLoopInfo*> topLevelOuterLoops = {};
```
Comment (kchabSS) [inline lib/core/sema/okl_sema_info.h:109]: Storing a list of small objects (pointers) is not worth it.
I would suggest making a method that selects all the `Outer` loops for each loop branch in the kernel.

File: lib/core/ast_processor_manager/ast_processor_manager.h [hunk 4]
```diff
@@ -120,6 +120,30 @@ HandleType makeSpecificSemaXXXHandle(Handler& handler) {
...
+        const auto localNode = clang::dyn_cast_or_null<ExprType>(&node);
+        if (!localNode) {
+            auto baseNodeTypeName = typeid(NodeType).name();
+            auto handleNodeTypeName = typeid(ExprType).name();
+            return tl::make_unexpected(
+                Error{{},
+                      util::fmt("Failed to cast {} to {}", baseNodeTypeName, handleNodeTypeName)
+                          .value()});
+        }
+        return handler(nullptr, *localNode, sema, stage);
```
Comment (kchabSS) [inline lib/core/ast_processor_manager/ast_processor_manager.h:143]: Always passing `nullptr` as `Attr` to the handler.
Please optimize.

File: lib/attributes/utils/cuda_subset/kernel.cpp [hunk 5]
```diff
@@ -71,7 +71,7 @@ HandleResult handleKernelAttribute(const Attr& a, const FunctionDecl& func, Sess
 
     auto startPos = getAttrFullSourceRange(a).getBegin();
     size_t n = 0;
-    for (auto& child : kernelInfo.children) {
+    for (auto* child : kernelInfo.topLevelOuterLoops) {
```
Comment (kchabSS) [inline lib/attributes/utils/cuda_subset/kernel.cpp:74]: The pointer is not validated.

File: lib/attributes/backend/launcher.cpp [hunk 6]
```diff
@@ -393,13 +393,13 @@ HandleResult handleLauncherKernelAttribute(const Attr& a,
     rewriter.ReplaceText(paramsRange, paramStr);
 
     size_t n = 0;
-    for (auto& loop : kernelInfo.children) {
-        removeAttribute(loop.attr, s);
+    for (auto* loop : kernelInfo.topLevelOuterLoops) {
```
Comment (kchabSS) [inline lib/attributes/backend/launcher.cpp:396]: The pointer is not validated.

File: lib/attributes/backend/dpcpp/kernel.cpp [hunk 7]
```diff
@@ -121,7 +121,7 @@ HandleResult handleKernelAttribute(const clang::Attr& a,
 
     size_t n = 0;
     auto startPos = getAttrFullSourceRange(a).getBegin();
-    for (auto& child : kernelInfo.children) {
+    for (auto* child : kernelInfo.topLevelOuterLoops) {
```
Comment (kchabSS) [inline lib/attributes/backend/dpcpp/kernel.cpp:124]: The pointer is not validated.

File: tests/functional/data/transpiler/backends/dpcpp/shared/shared_builtin_types_ref.json [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            "type": "custom"
+          },
+          "name": "range_",
+          "ptr": true
+        }
+      ],
+      "name": "_occa_function3_0"
+    }
+  ]
+}
```
Comment (kchabSS) [inline tests/functional/data/transpiler/backends/dpcpp/shared/shared_builtin_types_ref.json:89]: EOL at EOF

**Review by kchabSS — CHANGES_REQUESTED** [review]
Overall, looks OK. But some changes are questionable.

File: lib/core/sema/okl_sema_info.h (hunk 3 above)
Comment (pavlo-hilei) [reply lib/core/sema/okl_sema_info.h:109]: Do you mean to create a method that calculates this list from `topLevelLoops`?

Comment (pavlo-hilei) [reply lib/core/sema/okl_sema_info.h:109]: Why not worth it?

Comment (pavlo-hilei) [reply lib/core/sema/okl_sema_info.h:109]: Regarding question. We have to store them somewhere. You start to parse from top to bottom. First outer must have parent a regular (if there is a regular before it), So this regular must be stored after parsing it, and retrieved when parsing first outer

Comment (kchabSS) [reply lib/core/sema/okl_sema_info.h:109]: I see.
OK. I suggest to change `topLevelOuterLoops` to:
`std::vector<OklLoopInfo*> attributedChildren = {}`
Even if it is reallocated, we are storing raw pointers, so we are safe.

Comment (kchabSS) [reply lib/core/sema/okl_sema_info.h:109]: > Do you mean to create a method that calculates this list from `topLevelLoops`?

Yes. possibly just recursively visit all the child, and select first attributed in each branch.

Comment (pavlo-hilei) [reply lib/core/sema/okl_sema_info.h:109]: Alright

File: lib/core/ast_processors/okl_sema_processor/processor.cpp (hunk 1 above)
Comment (pavlo-hilei) [reply lib/core/ast_processors/okl_sema_processor/processor.cpp:36]: Good idea

File: lib/core/ast_processors/okl_sema_processor/processor.cpp (hunk 2 above)
Comment (pavlo-hilei) [reply lib/core/ast_processors/okl_sema_processor/processor.cpp:21]: +

File: lib/core/ast_processor_manager/ast_processor_manager.h (hunk 4 above)
Comment (pavlo-hilei) [reply lib/core/ast_processor_manager/ast_processor_manager.h:143]: Removing this argument will break a whole ton of code. I must add new DefaultStmtHandler and DefaultDeclHandler, then update makeDefaultHandler, then update type of default handler map, then AstProcessorManager::runPreActionNodeHandle and runNodeHandler assume that attribute and default handlers have the same interface, which is another ton of fixes.

Comment (kchabSS) [reply lib/core/ast_processor_manager/ast_processor_manager.h:143]: I see. well, it's OK.

File: tests/functional/data/transpiler/backends/dpcpp/shared/shared_builtin_types_ref.json (hunk 8 above)
Comment (pavlo-hilei) [reply tests/functional/data/transpiler/backends/dpcpp/shared/shared_builtin_types_ref.json]: Removed file (it was mistakenly pushed)

File: lib/attributes/backend/dpcpp/kernel.cpp (hunk 7 above)
Comment (pavlo-hilei) [reply lib/attributes/backend/dpcpp/kernel.cpp:124]: Added

File: lib/attributes/backend/launcher.cpp (hunk 6 above)
Comment (pavlo-hilei) [reply lib/attributes/backend/launcher.cpp:396]: Added

File: lib/attributes/utils/cuda_subset/kernel.cpp (hunk 5 above)
Comment (pavlo-hilei) [reply lib/attributes/utils/cuda_subset/kernel.cpp:74]: Added

**Review by kchabSS — APPROVED** [review]
Looks OK. Please validate all tests before merging.

**Pull request merged**